src/serac/numerics/functional/domain_integral.hpp

```diff
@@ -205,7 +205,7 @@ class DomainIntegral {
       }
     }
     if (which_local_trial_space != -1) {
-      action_of_gradient_[which_trial_space](input_E, output_E);
+      action_of_gradient_[which_local_trial_space](input_E, output_E);
```

Comment:
I don't fully understand the code yet (not because it's unclear, only because I haven't examined it before), though I can see the gist of the error. Maybe a unit test is in order? It's a trivial fix, and I don't want to hold it up. On the other hand, I could see someone breaking this in the future who isn't as familiar with it (like me). What do you think? Overkill for this?

Reply:
By unit test do you mean a test fixture for just `DomainIntegral` (rather than `Functional` or `SolidMechanics` module) that confirms this part is working? I guess we don't have proper unit tests for these classes in that sense, they almost all start at `Functional` scope or higher.

This conditional logic that changed is being tested through the physics modules, but we didn't have a test case that caught this issue.

Reply:
I should probably just change this implementation to be a little bit clearer though. Something like a `std::map` from the "global" argument index to the "local" index used by the `DomainIntegral` would work just as well and probably make more sense.

Reply:
Using a map sounds like a good idea. 

For the test suggestion, I'm applying a rule of thumb: if a user catches a bug that we didn't see in the code despite looking at it several times, it's probably the kind of misunderstanding that we could make again,  and so deserves a test.  I'm fine with the test being at another scope. I want to guard against a regression if someone refactors this code in the future, especially someone less familiar with its design.

I could be convinced that it isn't necessary in this case - I trust your judgment.

Reply:
Perhaps adding the situation that uncovered the bug (using a parameterized material with a parameterized body force) to one of the functional tests would be a good idea?

Reply:
I'm in the process of modifying one of the existing tests to cover this, and changing the implementation to (hopefully) make clearer what's going on with the index translation.